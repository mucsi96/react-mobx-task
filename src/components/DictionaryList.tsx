import * as React from 'react';
import { inject, observer } from 'mobx-react';
import DictionaryStore from '../stores/DictionaryStore';
import ViewStore from '../stores/ViewStore';

@inject('dictionaryStore', 'viewStore')
@observer
export default class DictionaryList extends React.Component {
  get injected() {
    return this.props as {
      dictionaryStore: DictionaryStore;
      viewStore: ViewStore;
    };
  }

  public render(): JSX.Element {
    if (!this.injected.dictionaryStore.dictionaries.length) {
      return (
        <div className="ui placeholder segment">
          <div className="ui header">No dictionaries yet.</div>
          {this.renderAddButton()}
        </div>
      );
    }

    return (
      <div className="ui segment">
        <h4 className="ui header">{'Dictionaries'}</h4>
        <div className="ui relaxed divided list">
          {this.injected.dictionaryStore.dictionaries.map(({ id, name }) => (
            <div key={id} className="item">
              <div className="content">
                <a className="header" onClick={() => this.handleClickEdit(id)}>
                  {name}
                </a>
              </div>
            </div>
          ))}
        </div>
        {this.renderAddButton()}
      </div>
    );
  }

  private renderAddButton() {
    return (
      <button
        autoFocus={true}
        className="ui primary button"
        onClick={this.handleClickAdd}
      >
        Add dictionary
      </button>
    );
  }

  private handleClickAdd = () => {
    this.injected.viewStore.createDictionary();
  };

  private handleClickEdit = (id: string) => {
    this.injected.viewStore.editDictionary(id);
  };
}
