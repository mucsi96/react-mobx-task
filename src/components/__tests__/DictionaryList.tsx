import * as React from 'react';
import * as RTL from 'react-testing-library';
import { Provider } from 'mobx-react';
import DictionaryList from '../DictionaryList';

const viewStore = {
  createDictionary: jest.fn(),
  showDictionary: jest.fn()
};

const dictionaryStore = {
  dictionaries: [{ id: '1', name: 'a' }, { id: '2', name: 'b' }],
  isDictionaryValid: (id: string) => (id === '1' ? true : false)
};

const emptyDictionaryStore = {
  ...dictionaryStore,
  dictionaries: []
};

const render = ({ empty = false } = {}) =>
  RTL.render(
    <Provider
      dictionaryStore={empty ? emptyDictionaryStore : dictionaryStore}
      viewStore={viewStore}
    >
      <DictionaryList />
    </Provider>
  );

describe('DictionaryList', () => {
  it('renders', () => {
    const { asFragment } = render();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders if there is no dictionary', () => {
    const { asFragment } = render({ empty: true });
    expect(asFragment()).toMatchSnapshot();
  });

  it('navigates to show dictionary view on item click', () => {
    viewStore.showDictionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('a'));
    expect(viewStore.showDictionary).toBeCalledWith('1');
  });

  it('navigates to create dictionary view on "Add dictionary" press', () => {
    viewStore.createDictionary.mockClear();
    const { getByText } = render();
    RTL.fireEvent.click(getByText('Add dictionary'));
    expect(viewStore.createDictionary).toBeCalled();
  });
});
