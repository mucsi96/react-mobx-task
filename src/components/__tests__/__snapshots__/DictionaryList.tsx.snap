// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DictionaryList renders 1`] = `
<DocumentFragment>
  <div
    class="ui basic segment"
  >
    <div
      class="ui header"
    >
      Dictionaries
    </div>
    <div
      class="ui relaxed divided list"
    >
      <div
        class="item"
      >
        <div
          class="content"
        >
          <a
            class="header"
          >
            a
          </a>
        </div>
      </div>
      <div
        class="item"
      >
        <div
          class="content"
        >
          <a
            class="header"
          >
            <i
              class="yellow exclamation triangle icon"
            />
            b
          </a>
        </div>
      </div>
    </div>
    <button
      class="ui primary button"
    >
      Add dictionary
    </button>
  </div>
</DocumentFragment>
`;

exports[`DictionaryList renders if there is no dictionary 1`] = `
<DocumentFragment>
  <div
    class="ui placeholder segment"
  >
    <div
      class="ui icon header"
    >
      No dictionaries yet.
    </div>
    <button
      class="ui primary button"
    >
      Add dictionary
    </button>
  </div>
</DocumentFragment>
`;
